File: backend/spring-boot/src/main/java/org/bootstrapbugz/backend/auth/jwt/event/email/JwtEmailSupplier.java
package org.bootstrapbugz.backend.auth.jwt.event.email;

import java.util.EnumMap;
import java.util.Map;
import java.util.function.Supplier;
import org.bootstrapbugz.backend.auth.jwt.util.JwtUtil.JwtPurpose;

public class JwtEmailSupplier {
  private static final Map<JwtPurpose, Supplier<JwtEmail>> emailType =
      new EnumMap<>(JwtPurpose.class);

  static {
    emailType.put(JwtPurpose.VERIFY_EMAIL_TOKEN, VerificationEmail::new);
    emailType.put(JwtPurpose.RESET_PASSWORD_TOKEN, ResetPasswordEmail::new);
  }

  public JwtEmail supplyEmail(JwtPurpose jwtPurpose) {
    final var emailSupplier = emailType.get(jwtPurpose);
    if (emailSupplier == null)
      throw new IllegalArgumentException("Invalid email type: " + jwtPurpose);
    return emailSupplier.get();
  }
}

File: backend/spring-boot-kotlin/src/main/kotlin/org/bootstrapbugz/api/auth/jwt/event/email/JwtEmailSupplier.kt
package org.bootstrapbugz.api.auth.jwt.event.email

import java.util.*
import org.bootstrapbugz.api.auth.jwt.util.JwtUtil.JwtPurpose

object JwtEmailSupplier {
  private val emailType: Map<JwtPurpose, () -> JwtEmail> =
    mapOf(
      JwtPurpose.VERIFY_EMAIL_TOKEN to ::VerificationEmail,
      JwtPurpose.RESET_PASSWORD_TOKEN to ::ResetPasswordEmail
    )

  fun supplyEmail(jwtPurpose: JwtPurpose): JwtEmail =
    emailType[jwtPurpose]?.invoke()
      ?: throw IllegalArgumentException("Invalid email type: $jwtPurpose")
}
